### ISSUE TITLE
[Regression in 11.0.16 / 17.0.4] libjawt.so: version `SUNWprivate_1.1' not found

### ISSUE THREAD

**Issue by sratz** [SAP/SapMachine#1227]
With SapMachine 11.0.16 it's currently not possible to use the SWT/AWT bridge, which worked fine in 11.0.15:

Steps to reproduce:
- Grab `libswt-awt-gtk-4954r6.so` from https://github.com/eclipse-platform/eclipse.platform.swt.binaries/tree/c4519af2c5f9d41f00bb8347f79b1aee0b30dbd4/bundles/org.eclipse.swt.gtk.linux.x86_64
- Compile and execute the snippet
   ```
   public class LinkErrorTest {
     public static void main(String[] args) {
       System.loadLibrary("jawt");
       System.load(args[0]);
       System.out.println("OK");
     }
  }
  ```

Results for different JVMs:

SapMachine 11.0.15:
```
$ ./java/sapmachine-jdk-11.0.15/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so
OK
```

Adoptium 11.0.16:
```
$ ./java/adoptium-jdk-11.0.16+8/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so
OK
```

SapMachine 11.0.16:
```
$ ./java/sapmachine-jdk-11.0.16/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so 
Exception in thread "main" java.lang.UnsatisfiedLinkError: /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so: /home/foo/java/sapmachine-jdk-11.0.16/lib/libjawt.so: version `SUNWprivate_1.1' not found (required by /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2501)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2700)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2630)
	at java.base/java.lang.Runtime.load0(Runtime.java:768)
	at java.base/java.lang.System.load(System.java:1837)
	at LinkErrorTest.main(LinkErrorTest.java:4)
```

SapMachine 11.0.17-ea.2:
```
$ ./java/sapmachine-jdk-11.0.17/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so
Exception in thread "main" java.lang.UnsatisfiedLinkErrorTest: /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so: /home/foo/java/sapmachine-jdk-11.0.17/lib/libjawt.so: version `SUNWprivate_1.1' not found (required by /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2501)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2700)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2630)
	at java.base/java.lang.Runtime.load0(Runtime.java:768)
	at java.base/java.lang.System.load(System.java:1837)
	at LinkErrorTest.main(LinkErrorTest.java:4)
```

This seems to be a regression within https://github.com/SAP/SapMachine/compare/sapmachine-11.0.15...sapmachine-11.0.16

**Comment by RealCLanger** [comment]
@MBaesken can you take a look?

**Comment by sratz** [comment]
I've tested with 17.0.x as well:

```
$ ./java/sapmachine-jdk-17.0.3/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so
OK
```
```
$ ./java/sapmachine-jdk-17.0.4/bin/java LinkErrorTest /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so
Exception in thread "main" java.lang.UnsatisfiedLinkErrorTest: /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so: /home/foo/java/sapmachine-jdk-17.0.4/lib/libjawt.so: version `SUNWprivate_1.1' not found (required by /home/foo/.swt/lib/linux/x86_64/libswt-awt-gtk-4954r6.so)
	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2389)
	at java.base/java.lang.Runtime.load0(Runtime.java:755)
	at java.base/java.lang.System.load(System.java:1953)
	at LinkErrorTest.main(LinkErrorTest.java:4)
```

**Comment by MBaesken** [comment]
We changed the builds ~ 2 months ago to a devkit, this might have changed the build results slightly.
Can be seen as well in the normal OpenJDK11 not only SapMachine 11 / 17.

**Comment by MBaesken** [comment]
Created on Ubuntu 20 an openjdk 11u_dev JVM ( sources from https://github.com/openjdk/jdk11u-dev ).
First with the machine provided tools :

/build_normal/images/jdk/bin/java  LinkErrorTest  /transfer2/sapmachine_linuxx64/libswt-awt-gtk-4954r5.so
OK

then (same sources) with the devkit x86_64-linux-gnu-to-x86_64-linux-gnu-fedora12-gcc8.3.0_new :

/build_devkit/images/jdk/bin/java  LinkErrorTest  /transfer2/sapmachine_linuxx64/libswt-awt-gtk-4954r5.so
Exception in thread "main" java.lang.UnsatisfiedLinkError: /transfer2/sapmachine_linuxx64/libswt-awt-gtk-4954r5.so: /build_devkit/images/jdk/lib/libjawt.so: version `SUNWprivate_1.1' not found (required by /transfer2/sapmachine_linuxx64/libswt-awt-gtk-4954r5.so)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2501)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2700)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2630)
	at java.base/java.lang.Runtime.load0(Runtime.java:768)
	at java.base/java.lang.System.load(System.java:1837)
	at LinkErrorTest.main(LinkErrorTest.java:4)

So the tooling in the devkit seems to introduce a slight difference.

Similar issue is here :  https://bugs.eclipse.org/bugs/show_bug.cgi?id=558874

**Comment by MBaesken** [comment]
there is a similar / related discussion in this thread 
https://mail.openjdk.org/pipermail/jdk-dev/2020-February/003865.html
especially
https://mail.openjdk.org/pipermail/jdk-dev/2020-February/003890.html
https://mail.openjdk.org/pipermail/jdk-dev/2020-February/003909.html

"but SUNWprivate_1.1 comes from the mapfile which was used in jdk8:
http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/make/mapfiles/libjawt/mapfile-vers
The mapfiles were removed in jdk11."

So my estimation is that the libswt-awt-gtk-4954r5.so (and r6) has been built using jdk8  JNI  and this way, it somehow got the SUNWprivate_1.1 dependency (when looking into the lib with objdump this seems to be the case).
But when running with jdk11 or jdk17 it is not always the case that the SUNWprivate_1.1 dependency  is still there (seems to rely on special compiler/linker etc. versions used).
We can try to find a 'good' toolchain that brings back the symbol but this is a bit tricky and  might break in the future again.

**Comment by MBaesken** [comment]
> We changed the builds ~ 2 months ago to a devkit, this might have changed the build results slightly.

The builds, after changing to the devkits , started using the gold-linker / binutils .
See  https://github.com/openjdk/jdk/blob/master/make/devkit/Tools.gmk#L433
--enable-gold=default \

This has slightly different behavior and is leading to the 'SUNWprivate_1.1 missing' issue described in this issue.

**Comment by RealCLanger** [comment]
We have modified the devkit with which we build SapMachine and the issue should be fixed with the latest update 11.0.16.1: https://github.com/SAP/SapMachine/releases/tag/sapmachine-11.0.16.1

Hence I'm closing this issue. Feel free to contact us again if you still find errors or need further help.

**Comment by RealCLanger** [comment]
PS: SapMachine 17 is fixed with 17.0.4.1: https://github.com/SAP/SapMachine/releases/tag/sapmachine-17.0.4.1
